pkg/image/admission/imagepolicy/rules/rules.go

```diff
@@ -0,0 +1,143 @@
+package rules
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/image/admission/imagepolicy/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type ImagePolicyAttributes struct {
+	Resource      unversioned.GroupResource
+	Name          imageapi.DockerImageReference
+	OriginalName  string
+	Image         *imageapi.Image
+	ExcludedRules sets.String
+}
+
+type RegistryMatcher interface {
+	Matches(name string) bool
+}
+
+type RegistryNameMatcher imageapi.DefaultRegistryFunc
+
+func (m RegistryNameMatcher) Matches(name string) bool {
+	current, ok := imageapi.DefaultRegistryFunc(m)()
+	if !ok {
+		return false
+	}
+	return current == name
+}
+
+type nameSet []string
+
+func (m nameSet) Matches(name string) bool {
+	for _, s := range m {
+		if s == name {
+			return true
+		}
+	}
+	return false
+}
+
+func NewRegistryMatcher(names []string) RegistryMatcher {
+	return nameSet(names)
+}
+
+type resourceSet map[unversioned.GroupResource]struct{}
+
+func (s resourceSet) addAll(other resourceSet) {
+	for k := range other {
+		s[k] = struct{}{}
+	}
+}
+
+func prepareSourceRule(rule *api.ImageCondition) (requiresImage bool, covers resourceSet) {
+	switch {
+	case len(rule.MatchImageLabels) > 0,
+		len(rule.MatchImageAnnotations) > 0,
+		len(rule.MatchSignatures) > 0,
+		len(rule.MatchDockerImageLabels) > 0:
+		requiresImage = true
+	}
+
+	covers = make(resourceSet)
+	for _, gr := range rule.OnResources {
+		covers[gr] = struct{}{}
+	}
+
+	return requiresImage, covers
+}
+
+// emptyImage is used when resolution failures occur but resolution failure is allowed
+var emptyImage = &imageapi.Image{
+	DockerImageMetadata: imageapi.DockerImage{
+		Config: &imageapi.DockerConfig{},
+	},
+}
+
+func matchSourceRule(rule *api.ImageCondition, integrated RegistryMatcher, attrs *ImagePolicyAttributes) bool {
+	if rule.MatchIntegratedRegistry && !integrated.Matches(attrs.Name.Registry) {
+		return false
+	}
+	if len(rule.MatchRegistries) > 0 && !hasAnyMatch(attrs.Name.Registry, rule.MatchRegistries) {
+		return false
+	}
+
+	// all subsequent calls require the image
+	image := attrs.Image
+	if image == nil {
+		if !rule.AllowResolutionFailure {
+			return false
+		}
+		// matches will be against an empty image
+		image = emptyImage
```

Comment:
> I suppose I could hoist resolution failures up and let their behavior be determined by the parent rule type. Not sure that's easier to reason about.

Pretty sure that `api.RequiresImage(rule)` will be false if all the big `if's` below are false and you return true.  Also, if any of the big if's below are true, you'll always return false since they're impossible to match on an empty image, right?

If that's not true, then the code here is so sticky I can't read it which worries me for other reasons.